### ISSUE TITLE
Handle Relogin differently than ExistingSessionDetected

### ISSUE THREAD

**Issue by jrc** [IbcAlpha/IBC#45]
In a headless scenario, it would be nice to force-disconnect any existing sessions that might still be running somewhere, possibly due to unclean termination for whatever reason.

Unfortunately, as currently implemented, `ExistingSessionDetectedAction` is used in both `ExistingSessionDetectedDialogHandler.java` and `ReloginDialogHandler.java`. Setting it to `primary` leads to the undesirable behavior documented in `config.ini`:

> Also note that if two primary sessions are started, they will both keep reconnecting and disconnecting the other: therefore care needs to be exercised in the use of this setting."

Could this be solved by splitting `ExistingSessionDetectedAction` into two different settings, so that one can set `ExistingSessionDetectedAction` to `primary` but leave, say, let's call it "ReloginDialogAction", to `manual` or `secondary`?

The desired effect is for a new instance to be able to force-disconnect any existing sessions without the other sessions fighting to re-login.

**Comment by rlktradewright** [comment]
What you describe is exactly what `ExistingSessionDetected=primary` is for. So you set your headless session to be primary, and you set any other session to be secondary. As the comment in `config.ini` suggests, you should never set primary for two setups that could possibly both be running at the same time.

Until a year or so ago, this was all perfectly straightforward, because the Relogin dialog only ever appeared after a new session had tried to take over, and only for the existing session (if I remember correctly).

However at that time, new behaviour appeared in TWS where the ReLogin dialog started appearing at other times (reason unknown), but there was no way for IBC to tell whether it was because a new conflicting session had started or because of this unknown reason: setting `ExistingSessionDetected=primary` covers this case, but only for a primary session. 

But for a secondary session, you would ideally want it to ok the ReLogin dialog if it occurred for this new reason, but not if it occurred because a primary session was starting. But since IBC can't distinguish between these two situations, having an additional setting wouldn't help.

So the upshot is that if you configure an IBC instance to be secondary (which is what you should do for any except your 'important' one), then it cannot automatically respond 'Ok' to the ReLogin dialog. This, I'm afraid, is just a fact of life, and I don't see any way around it. It has never been the slightest problem for me, and I don't really see why it should be a problem for anyone else.

**Comment by jrc** [comment]
Thanks for quick response (and for your good work on IBC). In our case, we don't have that primary/secondary distinction. We have only one configuration of IB Gateway that's running headless in the cloud, and we would like to be able to spawn a new server (of the same configuration) and have it reliably overtake any existing session(s), real or zombie. We try to terminate things cleanly but of course that's not always possible.

I wasn't aware of the subtleties of ReLogin, so if the above isn't feasible or makes little sense for most users, feel free to close this issue. I agree that in most cases, the desired behavior is already covered by simply setting `ExistingSessionDetected=primary`; I just wanted to avoid any possibility of that "note that if two primary sessions are started…" issue. Thanks.

**Comment by rlktradewright** [comment]
Ok, I see where you're coming from.. 

Are you aware that you can do a clean shutdown of IBC from the command line?

What this does is send a STOP command to IBC's command server via a TCP/IP connection, typically using `telnet`, and IBC then initiates a clean shutdown of TWS/Gateway as if the user had done a normal file exit. If you include this in your start script before running IBC, you could tell any known other instance to shut down before starting the new one.

To use the command server, you have to specifically configure it with a TCP port via which commands can be sent. See the section `IBC Command Server Settings` in `Config.ini`.

Note that because shutting down TWS/Gateway in this way closes the process, IBC goes with it, and therefore IBC is unable to report back when TWS/Gateway has completed its shutdown. So you'd have to address the issue of how to tell when the existing instance has gone, before starting the new one. 

One way would be to simply wait 'long enough' before starting the new instance: how long to delay could be determined by experiment, but of course there's always a possibility that some condition in TWS/Gateway might make it take longer occasionally. 

Another possibility is that because IBC does not close the command server's connection while processing a STOP command, there may be a way to catch the TCP/IP connection being broken as a result of the IBC process terminating: I can't advise on this as I haven't investigated it.

Yet another approach would be to write a simple program that connects to the command server, issues the STOP command, then monitors the socket for the connection breaking. Once this happens, exit the simple program, maybe wait another second or two just to be sure (this is probably paranoia speaking here!), then initiate the new IBC instance.

You would of course also have to handle the error that would occur if there is in fact no other IBC instance already running.

By the way, I assume you're running on Linux? If you're on Windows there is a Stop.bat command file included in the IBC download: this is included because for some bizarre reason, telnet on Windows won't read input from StdIn, so you can't script it easily to read the STOP command from a file: the included Stop.bat uses a trick that I found on StackOverflow to get round this.

**Comment by jrc** [comment]
>Are you aware that you can do a clean shutdown of IBC from the command line?

Well, the problem here really is only in the case where we didn't do a clean shutdown or some development instance is accidentally left running, which is the reason for needing this "override other sessions" functionality. Otherwise, we are catching `SIGTERM` and calling `EClient.disconnect()` and things are fine.

>By the way, I assume you're running on Linux?

Yep. (Docker)

Thanks.

**Comment by rlktradewright** [comment]
I must be missing something here.

The scenario you described is precisely why the `ExistingSessionDetectedAction` setting was provided in the first place. So you just make sure that all your development instances have `ExistingSessionDetectedAction=secondary`, and then the primary instance will always be able to start.

Is there some problem with this that I'm not seeing?

**Comment by jrc** [comment]
The issue isn't between development and production (we can use different accounts); it's handling multiple instances within development or within production. As mentioned, there's no concept of primary vs. secondary. The same configuration is being launched multiple times.

All of our instances are headless. `ExistingSessionDetectedAction=primary` almost does the right thing, except for that "note that if two primary sessions are started…" issue.

If the proposal doesn't make sense, feel free to close.

**Comment by rlktradewright** [comment]
Ok, thanks for the additional information.

What puzzles me is why you need to have multiple instances of TWS/Gateway running the same account at all.

For example, if you have multiple developers working on API application(s), why not just have a single TWS running on a server and have the applications connect to the API across the network? This is what I do all the time: although I'm only one person, I develop my trading platform and associated apps across multiple machines. Occasionally I need to do something on TWS (for example, close a position), so I remote into the server to access it: I can see this might be inconvenient with multiple developers, but much less so than the annoyances of having the developers run TWS instances locally.

But I suspect that what I just described isn't actually what you're talking about, in which case I still don't really understand your use case. Perhaps you're working on a product that will provide automated execution services for user-defined strategies, and you need to run up a Gateway intance for that user's IB account each time they use your service (if there isn't one already running)? That scenario could certainly suffer from the problem you're facing. 

By the way it not that 'the proposal doesn't make sense', it's just that if I fully understand your requirement I might be able to suggest something helpful: but the `ExistingSessionDetectedAction` setting is definitely not it!

Just one other idea to throw into the pot: it's perfectly possible to run IBC as a library rather than as a straight executable. This enables you to 'wrap' it with your own software to do whatever you like. In particular you could have some coordinating software that launches and keeps track of wrapped IBC instances, where the wrapper provides status information independent of whatever state TWS/Gateway is in. There is an example of using IBC in this mode in the [IBCLoader](https://github.com/IbcAlpha/IBC/tree/master/samples/IbcLoader) section of this repository.

**Comment by jrc** [comment]
Thanks for the response. I will try to do a better job of clarifying.

The use case is that we trade relatively infrequently, say once a day or so. So we spin up an IB instance, perform the trades (based on business logic from an external system) and grab the relevant data, and shut it down. This can happen frequently during development, but in production it happens in the cloud. As mentioned, everything is headless. So it's not the typical (I assume) use case where one runs IB continuously. It's more like the scenario you outlined ("automated execution services for user-defined strategies").

We considered running it a single instance as a server, as you suggested, but we didn't want to deal with the overhead in terms of admin and security, so for now we're not doing that, especially given the intended usage pattern.

**Comment by rlktradewright** [comment]
Ok, thanks again. That's a much clearer picture. 

So it sounds like the real problem that needs to be addressed is why the Gateway instance sometimes doesn't shut down. That then raises questions that your description doesn't cover:

* is the business logic that generates the trades in the same machine as the Gateway instance?

* does this business logic use the API to execute the trades, or does it use some sort of intermediary (such as a file which is then picked up by another program which itself uses the API)?

* how is the Gateway instance started?

* how is the Gateway instance stopped?

* are these 'orphaned' instances a result of some sort of failure, or just carelessness?

* is this a big deal, or just a minor annoyance? For example, could it happen in a 'live' situation? Is it actually worth spending much time trying to find a solution to this if it only happens infrequently, and only in a development environment?

**Comment by jrc** [comment]
In a Docker container, there's IB Gateway and a pile of Python code on top of it that vends a REST API used by the business logic, which lives in separate Docker container(s), which may or may not be on the same server. IB is started and stopped indirectly via container orchestration tools. 

This isn't a blocker or anything. We found a bug where we were not passing through SIGTERM from pid 1 from inside the Docker container, which may have been causing the problem, at least in some cases. But given that we need IB to start very reliably, I would still like a way to be able to virtually hit that "Continue Login" button if it ever appears, for whatever reason, even if it was due to some bug or carelessness.

**Comment by rlktradewright** [comment]
Ok, now all is crystal clear. Leave it with me. I think I can see how we might do this, but it's late on a Saturday night so I won't say any more about it now.

**Comment by rlktradewright** [comment]
I've made some modfications to IBC that will address this issue and give you what you need, and actually improve the handling of ExistingSessionDetected dialog all round.

I won't be able to test it till late this evening, because both my live and  my paper trading account are in use till then. 

The crucial observation that enables this, which I had never appreciated before, is that when you start a new instance when one is already running for the same account, the ExistingSessionDetected dialog always appears before the main window is created. But for the already running instance, the dialog appears after the main window is created. This enables me to distinguish between two situations. 

So I've introduced a new value of the `ExistingSessionDetectedAction` setting, namely `ExistingSessionDetectedAction=primaryoverride`. If the existing session and the new session are both 'primaryoverride', then the existing session will be stopped and the new one will proceed. However if the existing session is 'primary' then it will never yield (and this is now improved so that even if the new session is also 'primary', then the new session will be stopped, thus avoiding the switching back and forth that used to occur. Another benefit is that the 're-login' dialog is now ok'ed even if it occurs in a secondary session.

It's possible that I've overlooked something and when I get to test it, I'll find a flaw, but hopefully tomorrow I'll be able to give you a version to test.

**Comment by jrc** [comment]
Wow!! Great news. No rush from my side. Happy to test whenever you have something ready. Thanks again.

**Comment by rlktradewright** [comment]
Here's a link to a new beta version containing the updates. I've done some testing to ensure it basically works, but I'm hoping in your setup you can work it a bit harder.

[IBCLinux-3.8.0-beta.1](https://1drv.ms/u/s!AlqfLEOWDJ9Zh8JEMv_M_GJXyInRfQ)

You can just extract the `IBC.jar` and `version` files to replace your existing ones. The only other change is to `config.ini`, where you should read the comments relating to `ExistingSessionDetectedAction` (and perhaps update your existing `config.ini` with these comments).

Then change your setting in `config.ini` to:

`ExistingSessionDetectedAction=primaryoverride`

Now when you start a new IBC in a docker instance, any previous one for the same account should be terminated.

By the way, I'm not entirely sure 'primaryoverride' is the right name. It might be misunderstood as meaning that it will override a primary session, whereas it will only override another primaryoverride session. I'd prefer to use  'primaryoverrideable', but that's even longer. Any feelings either way (or something completely different)?

**Comment by jrc** [comment]
Thanks! I'll give it a spin.

`primaryoverrideable` (or simply `overrideable`) seems more self-explanatory and length here isn't a concern IMO.

**Comment by jrc** [comment]
Have not had time to work on this, but my colleague says regarding being logged in to the web-based IB Client Portal while trying to use TWS: "I noticed that if I just close the tab instead of logging out, my next login fails. only second attempt works." (I was not been able to reproduce this problem with my paper trading account.)

**Comment by rlktradewright** [comment]
Regarding `primaryoverrideable`, I agree that length per se isn't an issue, but there is a concern that the user might mis-spell it (eg, only one r, missing e after the d). What I might do is explicitly allow the obvious mis-spellings, and perhaps allow abbreviations: M, P, PO (or O) and S. `overrideable` on its own has its attractions, but losing its 'primary-ess' may be confusing.

**Comment by rlktradewright** [comment]
Using the Client Portal (or the mobile trader) while TWS Is running is asking for trouble, in my opinion.

I don't quite understand the description you quoted. What tab is he referring to: the client portal? What does he mean by 'the next login'? To what?

Also, what does he expect to happen? And what happens at TWS? And is TWS using `primaryoverrideable`?

It's not clear to me if there's something I need to spend time investigating here.

**Comment by jrc** [comment]
No worries. You've already been super-helpful and seems like you came up with a good solution. I will be traveling in the next week but I'll try to set aside some time to verify your fix.
